src/services/codefixes/fixAddMissingConstraint.ts

```diff
@@ -0,0 +1,66 @@
+/* @internal */
+namespace ts.codefix {
+    const fixId = "addMissingConstraint";
+    const errorCodes = [
+        // We want errors this could be attached to:
+        // Diagnostics.This_type_parameter_probably_needs_an_extends_0_constraint
+        Diagnostics.Type_0_is_not_comparable_to_type_1.code,
+        Diagnostics.Type_0_is_not_assignable_to_type_1_Two_different_types_with_this_name_exist_but_they_are_unrelated.code,
+        Diagnostics.Type_0_is_not_assignable_to_type_1_with_exactOptionalPropertyTypes_Colon_true_Consider_adding_undefined_to_the_types_of_the_target_s_properties.code,
+        Diagnostics.Type_0_is_not_assignable_to_type_1.code,
+        Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1_with_exactOptionalPropertyTypes_Colon_true_Consider_adding_undefined_to_the_types_of_the_target_s_properties.code,
+        Diagnostics.Property_0_is_incompatible_with_index_signature.code,
+        Diagnostics.Property_0_in_type_1_is_not_assignable_to_type_2.code,
+    ];
+    registerCodeFix({
+        errorCodes,
+        getCodeActions(context) {
+            const { sourceFile, span, program } = context;
+            const related = getDiagnosticRelatedInfo(program, sourceFile, span);
+            if (!related) {
+                return;
+            }
+            const changes = textChanges.ChangeTracker.with(context, t => addMissingConstraint(t, related));
+            return [createCodeFixAction(fixId, changes, Diagnostics.Add_extends_constraint, fixId, Diagnostics.Add_missing_new_operator_to_all_calls)];
+        },
+        fixIds: [fixId],
+        getAllCodeActions: context => codeFixAll(context, errorCodes, (changes, diag) => {
+            const info = getDiagnosticRelatedInfo(context.program, context.sourceFile, diag);
+            if (!info) return;
+            return addMissingConstraint(changes, info);
+        }),
+    });
+
+    function getDiagnosticRelatedInfo(program: Program, sourceFile: SourceFile, span: TextSpan) {
+        const diags = program.getSemanticDiagnostics(sourceFile).filter(diag => diag.start === span.start && diag.length === span.length);
+        if (!length(diags)) return;
+        const diag = diags[0];

```

Comment:
It sounds like you want `program.getSemanticDiagnostics(sourceFile).find(...)`

Reply:
Somewhat surprised that nobody has had to look at the diagnostics object for a code fix this whole time.